### PR TITLE
feat(select): autocomplete

### PR DESCRIPTION
**Pull request by noxify**
Theses changes add a new autocomplete option to the `select` prompt.

Inspired/based on: #62 

Closes: #92 

----

- [x] Update/Fix tests
- [x] Update docs

### CODE REVIEW

**Comment by noxify** [comment]
Tested it via the `demo.js` - It seems to work, but I have mentioned that the question isn't shown for when the answer is passed via the `PromptAgent`.

image

**Comment by noxify** [comment]
Also the tests are currently failing.

**Comment by noxify** [comment]
Fixed some problems - existing tests passed.

* Have to update the docs
* add missing tests for autocomplete
* Have to find the rootcause for the hidden question/answer if we use the `PromptAgent` in the demo - Any thoughts ?

**Comment by PierreDemailly** [comment]
> Have to find the rootcause for the hidden question/answer if we use the PromptAgent in the demo - Any thoughts ?

Not sure to understand, i can't reproduce locally. The question is shown with the answer as all others prompts.

On your screenshot [above](https://github.com/TopCli/prompts/pull/93#issuecomment-1938712634), the question+answer is also showed, you had 4 prompts in the demo.js (https://github.com/TopCli/prompts/blob/0ecdff3a6b5168120b69bdbcd32fa5940cf45807/demo.js) and I see 4 questions+answers in the screenshot. Am I missing something ?

**Comment by noxify** [comment]
fun fact: in the tests, the answer is shown. (test `should return the answer set via PromptAgent with autocomplete` )

**Comment by noxify** [comment]
Done - let me know if I have missed something.

File: test/select-prompt.test.ts [hunk 1]
```diff
@@ -406,4 +406,247 @@ describe("SelectPrompt", () => {
...
+      "? Choose between foo, bar or baz",
+      "› ",
+      " › foo",
+      "   bar",
+      "   baz",
+      // we press <b> so it filters values with 'b'
+      "› b",
+      " › bar",
+      "   baz",
+      // we press <b> so it filters values with 'ba'
```
Comment (PierreDemailly) [inline test/select-prompt.test.ts:465]:
```suggestion
      // we press <a> so it filters values with 'ba'
```

File: test/select-prompt.test.ts [hunk 2]
```diff
@@ -406,4 +406,247 @@ describe("SelectPrompt", () => {
...
+      "› ba",
+      " › bar",
+      "   baz",
+      // we press <return> so 'bar' is returned
+      "✔ Choose between foo, bar or baz › bar"
+    ]);
+    assert.deepEqual(input, "bar");
```
Comment (PierreDemailly) [inline test/select-prompt.test.ts:472]:
```suggestion
    assert.equal(input, "bar");
```

File: test/select-prompt.test.ts [hunk 3]
```diff
@@ -406,4 +406,247 @@ describe("SelectPrompt", () => {
...
+      "   baz",
+      // we press <backspace> so it filters all values
+      "› ",
+      " › foo",
+      "   bar",
+      "   baz",
+      // we press <return> so 'foo' is returned
+      "✔ Choose between foo, bar or baz › foo"
+    ]);
+    assert.deepEqual(input, "foo");
```
Comment (PierreDemailly) [inline test/select-prompt.test.ts:526]:
```suggestion
    assert.equal(input, "foo");
```

File: test/select-prompt.test.ts [hunk 4]
```diff
@@ -406,4 +406,247 @@ describe("SelectPrompt", () => {
...
+      "   bar",
+      "   baz",
+      // we press <B> so it filters values with 'b' or 'B' (case insensitive)
+      "› B",
+      " › bar",
+      "   baz",
+      // we press <return> so 'bar' is returned
+      "✔ Choose between foo, bar or baz › bar"
+    ]);
+    assert.deepEqual(input, "bar");
```
Comment (PierreDemailly) [inline test/select-prompt.test.ts:564]:
```suggestion
    assert.equal(input, "bar");
```

File: test/select-prompt.test.ts [hunk 5]
```diff
@@ -406,4 +406,247 @@ describe("SelectPrompt", () => {
...
+      "   bar",
+      "   baz",
+      // we press <b> so it filters value (case sensitive)
+      "› b",
+      " › bar",
+      "   baz",
+      // we press <return> so bar is returned
+      "✔ Choose between foo, bar or baz › bar"
+    ]);
+    assert.deepEqual(input, "bar");
```
Comment (PierreDemailly) [inline test/select-prompt.test.ts:613]:
```suggestion
    assert.equal(input, "bar");
```

File: test/select-prompt.test.ts [hunk 6]
```diff
@@ -406,4 +406,247 @@ describe("SelectPrompt", () => {
...
+      "› ",
+      " › foo",
+      "   bar",
+      "   baz",
+      // we press <B> so it filters no value (case sensitive)
+      "› B",
+      // we press <return> so bar is returned
+      "✖ Choose between foo, bar or baz › "
+    ]);
+    assert.deepEqual(input, "");
```
Comment (PierreDemailly) [inline test/select-prompt.test.ts:650]:
```suggestion
    assert.equal(input, "");
```

**Review by PierreDemailly — COMMENTED** [review]
lgtm %/ nits

**Review by fraxken — APPROVED** [review]

**Comment by PierreDemailly** [comment]
@all-contributors please add @noxify for code, test, doc

**Pull request merged**